stl/inc/mutex

```diff
@@ -45,11 +45,30 @@ public:
     _Mutex_base& operator=(const _Mutex_base&) = delete;
 
     void lock() {
-        _Check_C_return(_Mtx_lock(_Mymtx()));
+        if (_Mtx_lock(_Mymtx()) != _Thrd_success) {
+            // undefined behavior, only occurs for plain mutexes (N4928 [thread.mutex.requirements.mutex.general]/6)
+            _STD _Throw_Cpp_error(_RESOURCE_DEADLOCK_WOULD_OCCUR);

```

Comment:
We have tried to avoid throwing exceptions for undefined behavior - exceptions are for recovery, but UB should not be recovered from. I would argue that we should `_CSTD abort` when this precondition violation is detected. (Apparently SG1 thought otherwise, but if they really think an exception should be thrown, it should not be a precondition violation - similar to how various `basic_string` functions are required to throw `out_of_range`.)

Similarly below, although `try_lock()` doesn't throw, I think it would be better to abort than `return false;` (which transforms UB into guaranteed behavior). This is similar to the case of (IIRC) `deque::pop_back()`, where for a while we tolerated empty pops as no-ops.

That said, I could be convinced that this is not a regression in behavior (we were previously throwing a different exception / returning `false`) and we've lived with that for a decade.

Reply:
I see, you mentioned this in the PR description. I am okay with this then - at least it is very clearly commented. No change requested.

Reply:
> Similarly below, although `try_lock()` doesn't throw, I think it would be better to abort than `return false;` (which transforms UB into guaranteed behavior). This is similar to the case of (IIRC) `deque::pop_back()`, where for a while we tolerated empty pops as no-ops.

Oh, after checking it again I'm not sure whether the comment in `try_lock` is correct - `try_lock` seems able to fail for recursive mutex, in which case the behavior is still well-defined.